### ISSUE TITLE
Improve usage of `compiler-builtins`

### ISSUE THREAD

**Issue by jamesmunns** [rust-lang/rust#49851]
This is regarding the changes made in #49380. This change works, however when the user makes some kind of mistake, especially forgetting to install the new targets (like `thumbv7em-none-eabihf`), the error messages are not currently helpful.

The current error message when forgetting to install the target looks like this:

```bash
cargo build --target thumbv7em-none-eabihf --example blinky
   Compiling vcell v0.1.0
   Compiling bare-metal v0.1.1
   Compiling aligned v0.1.1
   Compiling untagged-option v0.1.1
error[E0463]: can't find crate for `compiler_builtins`

error[E0463]: can't find crate for `compiler_builtins`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error[E0463]: can't find crate for `compiler_builtins`

error: aborting due to previous error

error[E0463]: can't find crate for `compiler_builtins`
For more information about this error, try `rustc --explain E0463`.

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
```

compared to the following when you try to compile for a not-installed target:

```bash
cargo build --target i686-unknown-linux-gnu
   Compiling foo v0.1.0 (file:///tmp/foo)
error[E0463]: can't find crate for `std`
  |
  = note: the `i686-unknown-linux-gnu` target may not be installed

error: aborting due to previous error

For more information about this error
```

The following `#[no_std]` code shows (the first part of) a typical embedded project:

```rust
//! Blinks an LED

#![deny(unsafe_code)]
#![deny(warnings)]
#![no_std]

extern crate cortex_m;
```

using `cargo expand`, this code expands to:

```rust
#![feature(prelude_import)]
#![no_std]
//! Blinks an LED

#![deny(unsafe_code)]
#![deny(warnings)]
#![no_std]
#[prelude_import]
use core::prelude::v1::*;
#[macro_use]
extern crate compiler_builtins;
#[macro_use]
extern crate core;

extern crate cortex_m;
```

The Embedded WG talked about this, and we think the error messages would be improved if the order of `extern crate compiler_builtins;` and `extern crate core;` were swapped.

**Comment by jamesmunns** [comment]
CC @japaric and @oli-obk

**Comment by japaric** [comment]
> we think the error messages would be improved if the order of extern crate compiler_builtins; and extern crate core; were swapped.

I can confirm that swapping them produces the old, more helpful error messages of "note: the `thumbv6m-none-eabi` target may not be installed". It would also be good to drop the `#[macro_use]` from `extern crate compiler_builtins` as it seems like it could inject weird macros into `#![no_std]` programs.

cc @oli-obk

**Comment by jamesmunns** [comment]
Also I just noticed that in the expanded version, `#![no_std]` appears twice.

**Comment by oli-obk** [comment]
I wonder how that order swapping happens... should be easy enough to fix. The macro use comes from reusing the core import code. I guess we'll have to duplicate some now, but it's not too much

**Comment by oli-obk** [comment]
Ah.. inserting at position 0 will do that:

https://github.com/oli-obk/rust/blob/679657b863c2a53a3052d8af9defbce48e12db10/src/libsyntax/std_inject.rs#L61

This should be ann easy to fix issue. Also needs a ui test (just noting in case anyone tackles this)

**Comment by nagisa** [comment]
Furthermore, since `extern crate` statements can now appear at any position, it should be fine to replace that `insert` with a `push`.

**Comment by rcoh** [comment]
I have a diff for this but I'm having a hard time reproducing the error message in the `ui` tests. Any pointers @oli-obk ? I assume I need compiler flags but the only error I can seem to get is:
```
"Error loading target specification: Could not find specification for target \"thumbv7em-none-eabih\""
```

**Comment by rcoh** [comment]
Yep, that was it. Much appreciated.

**Comment by japaric** [comment]
This was fixed in #50006. Thanks @rcoh!
